pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+					k.waitNextMu.Unlock()
+					return false, false, nil
+				}
+				k.waitNextMu.Unlock()
+
+				currentDepth := k.NeighborhoodDepth()
+				if k.saturationFunc(po, currentDepth, k.connectedPeers) {
+					return false, false, nil
+				}
+
+				ma, err := k.addressBook.Get(peer)
+				if err != nil {
+					// either a peer is not known in the address book, in which case it
+					// should be removed, or that some severe I/O problem is at hand
+
+					if errors.Is(err, storage.ErrNotFound) {
+						k.logger.Errorf("failed to get address book entry for peer: %s", peer.String())
+						peerToRemove = peer
+						return false, false, errMissingAddressBookEntry
+					}
+
+					return false, false, err
+				}
+
+				k.logger.Debugf("kademlia dialing to peer %s", peer.String())
+				_, err = k.p2p.Connect(context.Background(), ma)
+				if err != nil {
+					k.logger.Debugf("error connecting to peer %s: %v", peer, err)
+					k.waitNextMu.Lock()
+					k.waitNext[peer.String()] = time.Now().Add(timeToRetry)
+					k.waitNextMu.Unlock()
+
+					// TODO: somehow keep track of attempts and at some point forget about the peer
+					return false, false, nil // dont stop, continue to next peer
+				}
+
+				k.connectedPeers.Add(peer, po)
+
+				k.waitNextMu.Lock()
+				delete(k.waitNext, peer.String())
+				k.waitNextMu.Unlock()
+
+				k.depthMu.Lock()
+				k.depth = k.recalcDepth()
+				k.depthMu.Unlock()
+
+				k.logger.Debugf("connected to peer: %s old depth: %d new depth: %d", peer, currentDepth, k.NeighborhoodDepth())
+
+				// the bin could be saturated or not, so a decision cannot
+				// be made before checking the next peer, so we iterate to next
+				return false, false, nil
+			})
+
+			if err != nil {
+				if errors.Is(err, errMissingAddressBookEntry) {
+					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
+					k.knownPeers.Remove(peerToRemove, po)
+				} else {
+					k.logger.Errorf("kademlia manage loop iterator: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// binSaturated indicates whether a certain bin is saturated or not.
+// when a bin is not saturated it means we would like to proactively
+// initiate connections to other peers in the bin.
+func binSaturated(bin, depth uint8, peers *pslice.PSlice) bool {
+	// short circuit for bins which are >= depth
+	if bin >= depth {
+		return false
+	}
+
+	// lets assume for now that the minimum number of peers in a bin
+	// would be 2, under which we would always want to connect to new peers
+	// obviously this should be replaced with a better optimization
+	// the iterator is used here since when we check if a bin is saturated,
+	// the plain number of size of bin might not suffice (for example for squared
+	// gaps measurement)
+
+	size := 0
+	_ = peers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		switch {
+		case po < bin:
+			return true, false, nil
+		case po > bin:
+			return false, true, nil
+		}
+
+		size++
+		return false, false, nil
+	})
+
+	return size >= 2
```

Comment:
Is this 2 the nnLowWatermark  or something different?

Reply:
well, no, the `nnLowWatermark` pertains only to the nearest neighborhood (hence, `nn...`). this is just some arbitrary low watermark i've put for the minimum number of peers that we'd want in a bin for it to constitute as saturated (or in other words that we are happy with this bin's condition and we don't want to dial out for new connections proactively), we can call this saturation or whatever we wanna call this later. i am also a bit confused by the definition of saturation (still, after 2.5 years).